src/endpoint.ts

```diff
@@ -1,26 +1,37 @@
-import { ArchethicRPCClient } from "./api/wallet_rpc.js";
+import { ConnectionState } from "./api/types.js";
+import { AWCWebBrowserExtension } from "./api/wallet_rpc.browserextension.js";
+import { ArchethicWalletClient } from "./api/wallet_rpc.js";
+import { AWCWebsocketStreamChannel } from "./api/wallet_rpc.websocket.js";
+
+export abstract class Endpoint {
+  abstract get isRpcAvailable(): boolean;
+  abstract get origin(): string;
+  abstract get nodeEndpoint(): URL | null;
 
-export class Endpoint {
   /**
-   * @param {String} endpoint
-   * @return {DirectEndpoint | WalletRPCEndpoint}
+   * @param {String | undefined} endpoint
+   * @return {Endpoint}
    */
-  static build(endpoint: string): DirectEndpoint | WalletRPCEndpoint {
-    const url: URL = new URL(endpoint);
-
-    if (url.protocol === "ws:") {
-      return new WalletRPCEndpoint(endpoint);
+  static build(endpoint: string | undefined): Endpoint {
```

Comment:
What do you think to support `String | AWCEndpoint | undefined` ? This way custom AWCStreamChannel might be passed as argument

Reply:
In my understanding, that method is a factory. Its role is to choose which kind of `Endpoint` to instanciate for a given endpoint `uri`.

I guess you suggest that change because once `AWCWebBrowserExtension` will be extracted from `libjs`, `build` method won't be able to build the `AWCEndpoint` anymore.

Maybe the solution would be to follow your next proposal 👇